keps/sig-network/2091-admin-network-policy/README.md

```diff
@@ -0,0 +1,1215 @@
+  L4Selector   AdminNetworkPolicyL4Selector
+
+  // List of sources from which traffic will be allowed/denied/passed to the entities 
+  // selected by this AdminNetworkPolicyRule. Items in this list are combined using a logical OR 
+  // operation. If this field is empty, this rule matches no sources. 
+  // If this field is present and contains at least one item, this rule
+  // allows/denies/passes traffic from the defined AdminNetworkPolicyPeer(s)
+  From         []AdminNetworkPolicyPeer
+}
+
+// AdminNetworkPolicyEgressRule describes an action to take on a particular 
+// set of traffic originating from pods selected by a AdminNetworkPolicy's 
+// Subject field. The traffic must match both l4Selector and to. 
+type AdminNetworkPolicyEgressRule struct {
+  // Name is an identifier for this rule. 
+  // +optional
+  Name         string
+
+  // Action specifies whether this rule must pass, allow or deny traffic. 
+  // Allow: allows the selected traffic 
+  // Deny: denies the selected traffic
+  // Pass: allows the selected traffic to skip and remaining positive priority AMP rules 
+  // and be delegated by K8's Network Policy. 
+  Action       AdminNetPolRuleAction
+
+  // L4Selector allows for matching on traffic based on L4 constructs. 
+  L4Selector   AdminNetworkPolicyL4Selector
+
+  // List of destinations to which traffic will be allowed/denied/passed from the entities 
+  // selected by this AdminNetworkPolicyRule. Items in this list are combined using a logical OR 
+  // operation. If this field is empty, this rule matches no destinations. 
+  // If this field is present and contains at least one item, this rule
+  // allows/denies/passes traffic to the defined AdminNetworkPolicyPeer(s)
+  To           []AdminNetworkPolicyPeer
+}
+
+// AdminNetworkPolicyL4Selector handles selection of traffic for based on L4 
+// constructs. Exactly one of the fields must be defined.
+type AdminNetworkPolicyL4Selector struct { 
+  // AllPorts cannot be "false" when it is set 
+  // AllPorts allows the user to select all ports for all protocols, thus not 
+  // selecting traffic based on L4 principles.
+  // If "true" then all ports are selected for the all protocols
+  AllPorts   *bool
+
+  // List of ports for outgoing traffic.
+  // Each item in this list is combined using a logical OR. If this field is
+  // empty or missing, this rule matches no ports (traffic not restricted by port).
+  // If this field is present and contains at least one item, then this rule 
+  // allows/denies/passes traffic only if the traffic matches at least one port in the list.
+  // +optional
+  Ports      []AdminNetworkPolicyPort 
+}
+
+
+// AdminNetworkPolicyPort describes a port to select
+type AdminNetworkPolicyPort struct {
+  // The protocol (TCP, UDP, or SCTP) which traffic must match. If not specified, this
+  // field defaults to TCP.
+  // +optional
+  Protocol   *v1.Protocol
+
+  // The port on the given protocol. This can either be a numerical or named
```

Comment:
Careful - adding named ports here can get very expensive in terms of rule explosion. That's because, theoretically, every pod could expose the same name via a different port. I know @danwinship has some opinions about named ports in this context.

Reply:
I thought about saying no to this.  But named ports exist and are a thing that Service can use.  It seems wrong to not support it here...  Open to being convinced

Reply:
I'll try to convince you:
1. it forces the implementation to do the mapping between port names and numbers, without a name, is just a number, it saves time ..
2. as Casey say, a named port can mean something different in the same rule
2. the combination with EndPort _This field cannot be defined if ... not defined .... or if the port field is defined as a named (string) port_ is :grimacing: 
3. I hear you several times to regret the decision on using the `IntOrString` type ;)

Reply:
> I know @danwinship has some opinions about named ports in this context.

My opinion is that we should have removed them from NP before v1 because no one used them and no one implemented them and they block certain optimizations in the backend.

But (a) most plugins eventually implemented it, and (b) NP has lots of features that are hard to optimize.

If they're part of NP, they should most likely be part of ANP as well.